CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md

```diff
@@ -0,0 +1,79 @@
+* loss of basic associativity and loss of other basic properties of real arithmetic, and
+* catastrophic cancellation.
+
+In addition, the author covers IEEE floating-point standards including the IEEE 754 binary standard which addresses:
+
+* bit-wise floating-point number data-structures (i.e. significant digits, exponents, sign),
+* rounding and accuracy requirements for basic floating-point operations,
+* overflow, underflow and special quantities (i.e. (signed) zero, demormalized numbers, NaNs, (signed) infinity &infin;),
+* floating-point exceptions, flags, and trap handlers,
+* rounding modes, and
+* conversions with integers and decimal (i.e. human-readable text-based) formats.
+
+Great attention is given to the issue of catastrophic cancellation (the subtraction of two nearby numbers where the operands are also subject to rounding errors) and related floating-point issues.
+Several examples are given of how different numerical expressions can be reformulated to eliminate a catastrophic cancellation and/or avoid overflow/overflow and thereby increase the robustness and accuracy of the floating-point calculation to a wide range of input data.
+(Reformulating an expression to reduce floating-point errors and exceptions can often increase the complexity of the code and increase the operation count and computational cost so there are often trade-offs of numerical accuracy, robustness, code maintainability, and runtime speed that must be made when implementing a numerical algorithm using floating-point computations.)
+
+System and compiler issues are also discussed where violations of basic floating-point foundations can make it very difficult to implement a robust numerical algorithm.
+For example, the author unequivocally states:
+
+> A language definition that does not require parentheses to be honored is useless for floating-point calculations.
+
+Several compilers and systems over the years have been known to violate this and other necessary basic foundations of floating-point and the IEEE 754 standard, especially when compiler optimizations are turned up.
+In addition, full implementation of the IEEE 754 standard on a system is rare and controlling behavior such as floating-point exceptions, rounding modes, and trap handlers is difficult or impossible in a portable way with many programming languages and libraries.
+This is especially true on some of the more bleeding-edge high-performance computing (HPC) accelerators and systems.
+As a result, numerical algorithm developers often have to rely on a smaller subset of the IEEE standard in order to develop a maximally portable robust implementation.
+
+Finally, the paper contains quite a few proofs of floating-point arithmetic and these proofs serve several purposes.
+First, the proofs establish the foundation of floating-point operations which demonstrate that floating-point is not black magic but is a well established mathematical approximation/discritization of real number arithmetic.
```

Comment:
discretization

Reply:
Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.